### ISSUE TITLE
Viewing logs of specific path in large repo is much slower compared to git

### ISSUE THREAD

**Issue by azuwis** [jj-vcs/jj#4674]
## Description

For example, viewing logs of `README` in linux kernel repo, git takes about 0.9s, jj takes 68s, 70 times slower.

## Steps to Reproduce the Problem

```
$ git clone https://github.com/torvalds/linux.git linux-git
$ cd linux-git
$ time git log -- README >/dev/null

real    0m0.936s
user    0m0.732s
sys     0m0.204s
```

```
$ jj git clone https://github.com/torvalds/linux.git linux-jj
$ cd linux-jj
$ time jj log --no-graph README >/dev/null

real    1m8.155s
user    1m8.037s
sys     0m0.597s
```

## Expected Behavior

The performance of viewing specific path in large repo should be comparable to git.

## Actual Behavior

See above.

## Specifications

- Platform: Linux
- Version: 0.22

**Comment by joyously** [comment]
Did you try timing the second time you do a `jj log` command? (Is the first one doing imports from Git for the whole repo or would that be done on the clone command?)

**Comment by emilazy** [comment]
IIRC from discussing on the Discord in the past there are two main contributors to this:

1. The default log template notes whether commits are empty; for merge commits, this involves merging the parents and checking if there is any difference (e.g. because of conflicts being resolved), and there is currently no caching of this. You can customize the templates (potentially per–repository) to avoid this.

2. Git does fancy smart bloom filter stuff that I don‘t understand to skip over commits when searching the log like that. Jujutsu doesn‘t do any of that at present. Git also has slightly different behaviour: merge commits are never shown at all in `git log ‹path›`.

**Comment by joyously** [comment]
The `sys time`s of both are very similar. It's the `real time` that is inflated.

**Comment by joyously** [comment]
When I try it, I get 
> Warning: The argument "README" is being interpreted as a path. To specify a revset, pass -r "README" instead.

Does that indicate a different logic path?

**Comment by arxanas** [comment]
Are you using commit-graph https://git-scm.com/docs/commit-graph? Can you try with commit graph disabled? (I think `git -c core.commitGraph=false log ...`)

**Comment by martinvonz** [comment]
> 1. You can customize the templates (potentially per–repository) to avoid this.

That's not actually enough because we still do the same calculation to even tell whether to include a merge commit in the result. I think Git instead compares the merge commit against each of its parents to see if the file is modified there. I suppose we should be able to do that as a first step and only do the merge if more than one diff contains the file.

**Comment by yuja** [comment]
Git appears to omit merged branches if one of the parents has identical content. That's probably the main factor why `git log README` is super fast in (merge-heavy) Linux repo.
https://github.com/git/git/blob/v2.47.0/revision.c#L1200

`git -c core.commitGraph=false log --full-history README` is still faster than jj, but the performance is closer.

**Comment by martinvonz** [comment]
> Git appears to omit merged branches if one of the parents has identical content.

Good idea!

> `git -c core.commitGraph=false log --full-history README` is still faster than jj, but the performance is closer.

I can confirm: On my machine, that takes about 17 seconds instead of just under 1 second without `--full-history`.

**Comment by pluiedev** [comment]
How would one approach fixing this? I've been using jj daily on Nixpkgs for about 2 weeks now and the speed is by far one of only two things that's been holding me back from recommending it to everyone around me 😅

**Comment by yuja** [comment]
Perhaps, the conceptually simplest option is to add an index of file paths that touch commits. Git builds a bloom filter for that purpose, but there might be a better option.

In order to implement `git log` (without `--full-history`) behavior, we'll need to add a revset function that queries commits by file paths but omits "uninteresting" branches. Maybe this revset function can leverage graph iterator internally, but we'll need some modifications on it.

Another related issue is https://github.com/martinvonz/jj/issues/4579. I think `jj log -r 'ancestors(.., nth_parent=1)' PATH` is equivalent to `git log --first-parent PATH`, which should be fast in merge-heavy repos. This can't replace all use cases of `jj log PATH`, but it's relatively easy to implement.

**Comment by emilazy** [comment]
Would just caching merge results somehow make a meaningful difference? Nixpkgs is indeed very merge‐heavy and the vast majority of those merges are empty.

**Comment by yuja** [comment]
> Would just caching merge results somehow make a meaningful difference?

If `-r~merges() PATH` is fast enough, yeah, it will save time. I'm not sure if that's the case for nix. To address the original problem, we'll need to avoid unpacking commit/tree objects as much as possible.

**Comment by emilazy** [comment]
It seems better, but unfortunately still unusably slow compared to `git log`, so I suppose the index is the best approach. That would allow keeping the current behaviour of including merge commits too, right?

**Comment by arxanas** [comment]
> @joyously: When I try it, I get
> 
> > Warning: The argument "README" is being interpreted as a path. To specify a revset, pass -r "README" instead.
> 
> Does that indicate a different logic path?

@joyously btw that warning is spurious; we actually do want to interpret `README` as a path and not a revset, so your test was doing the right thing.

---

> @yuja: Perhaps, the conceptually simplest option is to add an index of file paths that touch commits. Git builds a bloom filter for that purpose, but there might be a better option.

Before computing a full diff for a commit, git-branchless will first perform set intersection on the set of changed paths for a commit, which is a tree object comparison rather than full diffing. This lets it skip most uninteresting commits for certain operations (like deduplication).
- Does jj already do this? Is it computing a full diff for these cases? If so, maybe it's worth a try to add that fast path before building a persistent index?
- Also, git-branchless will do this in parallel; not sure if jj behaves differently.
- This also generalizes the behavior of skipping commits which are unchanged with respect to a certain parent... or maybe not, given that @martinvonz's statement that jj still tries to redo the merge.
- Possibly this suggestion isn't relevant if `jj log` is primarily slow due to merge commits. But @emilazy suggests that it's still "unusably slow compared to `git log`" even when excluding merges, so maybe it's relevant?

---

I know that Sapling has a specific system/service to serve blame/annotate queries, which would presumably be relevant for the case of following the log for a specific file.

- I glanced through the [public docs](https://sapling-scm.com/docs/category/internals) and the [`derived_data/` directory](https://github.com/facebook/sapling/tree/08b596d8e402d122a121ea16a01c7ce5ed2a3aad/eden/mononoke/derived_data), but didn't see anything obvious to suggest that e.g. there's an index keyed by file path rather than commit, which would be architecturally interesting.
- To be clear, I think everyone is proposing an index with the commit as the key and the set of changed paths as the value? I just wanted to see if Sapling did it differently.

For reference, here's some documentation on the "bloom filter" idea mentioned for Git: https://github.com/git/git/blob/25b0f41288718625b18495de23cc066394c09a92/Documentation/gitformat-commit-graph.txt#L136-L164.
- All commit-graph data is indexed with the commit as the key.
- I don't know if people are really using the bloom filter feature in practice.
- Interestingly, changes to more than 512 paths have all bits set to `1`, which seems like it might not be as useful at scale for merge commits?

**Comment by emilazy** [comment]
Would profiling traces be useful, by the way? Although anyone can generate them just by poking around with Jujutsu on Nixpkgs :)

**Comment by martinvonz** [comment]
> Perhaps, the conceptually simplest option is to add an index of file paths that touch commits. Git builds a bloom filter for that purpose, but there might be a better option.

Yes, I think that's what we'll have to do, but it's clearly a bit of work to implement.

> * Does jj already do this? Is it computing a full diff for these cases? If so, maybe it's worth a try to add that fast path before building a persistent index?

Yes, we already don't attempt to merge file contents in files that are not matched against. The code is here:

https://github.com/martinvonz/jj/blob/a6c18e8353d12085abe5e03f210987430f4ebd53/lib/src/default_index/revset_engine.rs#L1240-L1275

Because we pass in the file matcher into `diff_stream`, we won't call `resolve_file_values()` for files that don't match.

> * Also, git-branchless will do this in parallel; not sure if jj behaves differently.

Do you mean that git-branchless does it in parallel for multiple commits? How many commits ahead?

> * Possibly this suggestion isn't relevant if `jj log` is primarily slow due to merge commits. But @emilazy suggests that it's still "unusably slow compared to `git log`" even when excluding merges, so maybe it's relevant?

I'm a bit surprised that jj is so much slower than Git when there are no merges and Git's commit graph is disabled. I can't find it now but I think I saw some say that. I don't know why that would be. Profiling might help.

**Comment by emilazy** [comment]
FWIW:

```
yuyuko:~/Developer/nixpkgs  (empty) (no description set)
❭ hyperfine --warmup 2 'jj --ignore-working-copy --no-graph -T builtin_log_oneline -r "::@ ~ merges()" pkgs/stdenv/linux/default.nix | head -n 25' 'git log --oneline pkgs/stdenv/linux/default.nix | head -n 25'
Benchmark 1: jj --ignore-working-copy --no-graph -T builtin_log_oneline -r "::@ ~ merges()" pkgs/stdenv/linux/default.nix | head -n 25
  Time (mean ± σ):     11.447 s ±  0.142 s    [User: 10.777 s, System: 0.637 s]
  Range (min … max):   11.302 s … 11.734 s    10 runs

Benchmark 2: git log --oneline pkgs/stdenv/linux/default.nix | head -n 25
  Time (mean ± σ):      2.138 s ±  0.042 s    [User: 1.960 s, System: 0.179 s]
  Range (min … max):    2.090 s …  2.201 s    10 runs

Summary
  git log --oneline pkgs/stdenv/linux/default.nix | head -n 25 ran
    5.35 ± 0.12 times faster than jj --ignore-working-copy --no-graph -T builtin_log_oneline -r "::@ ~ merges()" pkgs/stdenv/linux/default.nix | head -n 25
```

So I would say it is still pretty dire even if you skip merges, assuming my methodology here is reasonable.

**Comment by emilazy** [comment]
`-T "commit_id ++ \"\n\""` seems to be comparably slow (actually… several seconds slower? But I don’t know if that’s just my ambient system load or thermals or whatever).

Turning off the commit graph narrows the gap to Jujutsu being “only” twice as slow:

```
yuyuko:~/Developer/nixpkgs  128  (empty) (no description set)
❭ hyperfine --warmup 2 'git -c core.commitGraph=false log --oneline pkgs/stdenv/linux/default.nix | head -n 50' 'git -c core.commitGraph=true log --oneline pkgs/stdenv/linux/default.nix | head -n 50'
Benchmark 1: git -c core.commitGraph=false log --oneline pkgs/stdenv/linux/default.nix | head -n 50
  Time (mean ± σ):      5.380 s ±  0.077 s    [User: 5.101 s, System: 0.269 s]
  Range (min … max):    5.291 s …  5.517 s    10 runs

Benchmark 2: git -c core.commitGraph=true log --oneline pkgs/stdenv/linux/default.nix | head -n 50
  Time (mean ± σ):      2.877 s ±  0.047 s    [User: 2.653 s, System: 0.224 s]
  Range (min … max):    2.790 s …  2.934 s    10 runs

Summary
  git -c core.commitGraph=true log --oneline pkgs/stdenv/linux/default.nix | head -n 50 ran
    1.87 ± 0.04 times faster than git -c core.commitGraph=false log --oneline pkgs/stdenv/linux/default.nix | head -n 50
```

So it’s probably a pretty big deal here, although that still leaves a lot of performance gap left to be explained.

**Comment by martinvonz** [comment]
> Turning off the commit graph narrows the gap to Jujutsu being “only” twice as slow:

Maybe most of the remainder is what Yuya mentioned in https://github.com/martinvonz/jj/issues/4674#issuecomment-2423476227

**Comment by emilazy** [comment]
Doesn’t my use of `::@ ~ merges()` rule that out?

**Comment by martinvonz** [comment]
Correct. We could skip entire branches the file has not changed on (but we risk missing cases where it was changed and then reverted).

**Comment by emilazy** [comment]
I suppose I’d prefer focusing on things like indexing and caching rather than things that might change semantics. (Not that I’m the one doing the work.)

**Comment by emilazy** [comment]
FWIW, there are less than 25, so the `| head`s there will not stop the entire history from being scanned. The file also just almost never changed. That definitely accentuates the problems here – 100× slower is scary! – but I just wanted ot point out that it does mean it reflects less well the common use‐case of wanting to see one pager screen’s worth of log output quickly for a file that has been changed at least semi‐recently.

**Comment by martinvonz** [comment]
Yes, I think it's more fair to compare the time it takes to produce the full output. Because the `jj log` output doesn't include merge commits, it needs to walk further. The `git log` case finds some merge commits and can therefore terminate the walk earlier.

**Comment by martinvonz** [comment]
So, in summary there are many different reasons for the differences:

1. `jj log` filters merge commits by comparing the auto-merged parents to the contents in (i.e. after) the commit. I think this contributes most to the difference. Use `~ merges()` in the revset and pass `--no-merges` to Git to make it more fair.
2. Because of (1), `jj log` has to walk further to produce the same amount of output (which I think is a good thing), so comparing how long it takes to produce a certain amount of output is unfair.
3.  Without `--full-history`, `git log` skips parts of the history that didn't contribute to the result
4. Git's commit graph has a Bloom filter that speeds up filtering by paths
5. Git has more low-level optimizations. @yuja has made some recent improvements

The above is just an explanation, not an excuse - we can and should still work to improve performance.

**Comment by arxanas** [comment]
> @martinvonz: Yes, we already don't attempt to merge file contents in files that are not matched against. The code is here: [...] Because we pass in the file matcher into diff_stream, we won't call resolve_file_values() for files that don't match.

If I understand the code correctly, there's no middle ground between "doing the fast path for zero changes" and "starting to calculate the full diff". Is that correct? I'm proposing that we determine and compare *all* changed paths before computing *any* content-based diffs.

- I don't think I understand the significance of `resolve_file_values` for the general case of following a path in `jj log`.
  - It sounds like it's specific to resolving conflicts for the cases of merge commits only, not generally computing a diff. Is that right?
  - Is it the case that, by this point/for each file in the loop, we're computing or have already computed the diff for that file? (Hopefully we would be able to avoid computing diffs for most commits in this case.)
- The issue with git-branchless + libgit2 might have been more severe than for the native jj diff implementation, though.
  - IIRC libgit2 does an O(n) pass over all (recursive) entries on both sides, rather than avoiding descending into unchanged trees, because it reuses an index-diffing code path, and it just has an iterator adapter for traversing an entire tree as if it were an index, or something like that.
- The code you linked calls `.diff_stream` which returns `BoxStream<'_, CopiesTreeDiffEntry>`. Is it possible that we're doing copy/rename detection now, and that makes it very slow?
  - Is there a quick way for someone to test it without copy/rename detection if they wanted?
  - IIUC, detecting copies/renames shouldn't affect whether the commit is returned for `jj log` on a file path. It should only affect the actual diff (and possibly other matchers, like those which check diff content). Is that correct?
- Perhaps the current `Matcher` system is a little too dynamic to implement optimizations like computing changed file paths before the diffs themselves, but I don't know.

> @martinvonz: Do you mean that git-branchless does it in parallel for multiple commits? How many commits ahead?

Yes, but git-branchless does it for commit deduplication and doesn't "stream" any results like we might expect `jj log` to.
- The parallelization is on top of whatever `rayon`'s default global thread pool is, which is probably ~1 thread per CPU.
- I don't specifically try to schedule any commits to happen earlier than others (possibly what you're referring to with "how many commits ahead?", since we probably want to roughly match the graph walk order).

The following stages are performed with parallelization within each stage only:

1. For each commit to be rebased + each "upstream" commit to be rebased onto, it computes a changed-path set for that commit.
1. For each upstream commit, it checks to see if its changed-path set matches any source changed-path set exactly. If not, then it's filtered out.
1. For each remaining upstream commit, its "patch ID" is computed. The results are added to a set of upstream patch IDs.
1. For each source commit, its patch ID is computed. If the resulting patch ID is in the upstream patch ID set, then that commit is marked to be skipped instead of rebased.

The above can be several thousands of upstream commits, so parallelizing it was important for performance. But it also has more "batch" characteristics than the `jj log` case, since it's not important to get results for some parts of the graph earlier.

Is jj's graph walk/filtering already done in parallel? (Is Git's?)

**Comment by yuja** [comment]
> If I understand the code correctly, there's no middle ground between "doing the fast path for zero changes" and "starting to calculate the full diff". Is that correct? I'm proposing that we determine and compare _all_ changed paths before computing _any_ content-based diffs.
> 
>  * I don't think I understand the significance of `resolve_file_values` for the general case of following a path in `jj log`.
>       
>    * It sounds like it's specific to resolving conflicts for the cases of merge commits only, not generally computing a diff. Is that right?

Correct. Here content-diffing is only needed when merging file conflicts.

Paths not matching `matcher` are excluded (so never attempted to be merged), but it's true that we can avoid content-diffing/merging if there are any other non-conflicting files that differ based on IDs. We don't have such optimization right now.

> `.diff_stream` which returns `BoxStream<'_, CopiesTreeDiffEntry>`. Is it possible that we're doing copy/rename detection now, and that makes it very slow?

No, at least for now. `files()` query doesn't support copy tracking yet.

---

FWIW, with quick testing, computing `has_diff_from_parent()` in parallel across multiple commits makes `log PATH` query 2x-3x faster. It might be even faster if we do things right, but we'll need non-trivial amount of work to parallelize the revset engine internals. Maybe better to try other low-hanging fruit first.

**Comment by patmaddox** [comment]
I am using jj on [freebsd/freebsd-src](https://github.com/freebsd/freebsd-src) and it's really slow. `jj log share/mk/suite.test.mk -n 1` takes `94.61 real`.

Adding `--no-graph` speeds it up, when the the latest commit is recent. My most recent commit is to that path and it is `0.08 real`. However, `-n 2` takes `4.96 real`.

For comparison, `time git log share/mk/suite.test.mk` takes `1.17 real`.

It's not clear to me if there are other flags / templates I can use to speed it up. For now, I can't really use `jj log` on this repo.

**Comment by dotdash** [comment]
Simply iterating over all parents in `has_diff_from_parent` and checking if any of them has no diff reduces the runtime for the `README` log in the Linux repo by about 40%:
```
~/src/linux-jj 🥋 zxky | (empty) (no description set) 
❯ hyperfine --warmup 1 --min-runs 2 'jj log --ignore-working-copy --no-graph -T builtin_log_oneline -r "::@" README'
Benchmark 1: jj log --ignore-working-copy --no-graph -T builtin_log_oneline -r "::@" README
  Time (mean ± σ):     45.945 s ±  0.853 s    [User: 44.129 s, System: 1.804 s]
  Range (min … max):   45.341 s … 46.548 s    2 runs
 
~/src/linux-jj 🥋 zxky | (empty) (no description set) 
❯ hyperfine --warmup 1 --min-runs 2 'jj log --ignore-working-copy --no-graph -T builtin_log_oneline -r "::@" README'
Benchmark 1: jj log --ignore-working-copy --no-graph -T builtin_log_oneline -r "::@" README
  Time (mean ± σ):     27.026 s ±  0.074 s    [User: 25.327 s, System: 1.694 s]
  Range (min … max):   26.974 s … 27.078 s    2 runs
```

Most savings come from not having to calculate the common ancestors for the merges.

The next step would be to also skip the uninteresting commits in the history walking code, but I didn't quite understand that code yet, so I couldn't test that.

FWIW, this is the quick hack I used for the above timings:
```diff
diff --git a/lib/src/default_index/revset_engine.rs b/lib/src/default_index/revset_engine.rs
index 9d8c79a6d0..e6edc04c65 100644
--- a/lib/src/default_index/revset_engine.rs
+++ b/lib/src/default_index/revset_engine.rs
@@ -55,6 +55,7 @@
 use crate::matchers::Visit;
 use crate::merge::Merge;
 use crate::merged_tree::resolve_file_values;
+use crate::merged_tree::MergedTree;
 use crate::object_id::ObjectId as _;
 use crate::repo_path::RepoPath;
 use crate::revset::ResolvedExpression;
@@ -1284,24 +1285,33 @@
         }
     }
 
-    // Conflict resolution is expensive, try that only for matched files.
-    let from_tree = rewrite::merge_commit_trees_no_resolve_without_repo(store, &index, &parents)?;
     let to_tree = commit.tree()?;
-    // TODO: handle copy tracking
-    let mut tree_diff = from_tree.diff_stream(&to_tree, matcher);
-    async {
-        // TODO: Resolve values concurrently
-        while let Some(entry) = tree_diff.next().await {
-            let (from_value, to_value) = entry.values?;
-            let from_value = resolve_file_values(store, &entry.path, from_value).await?;
-            if from_value == to_value {
-                continue;
+    let has_diff = |from_tree: MergedTree| {
+        // TODO: handle copy tracking
+        let mut tree_diff = from_tree.diff_stream(&to_tree, matcher);
+        async {
+            // TODO: Resolve values concurrently
+            while let Some(entry) = tree_diff.next().await {
+                let (from_value, to_value) = entry.values?;
+                let from_value = resolve_file_values(store, &entry.path, from_value).await?;
+                if from_value == to_value {
+                    continue;
+                }
+                return Ok(true);
             }
-            return Ok(true);
-        }
-        Ok(false)
+            Ok(false)
+        }
+        .block_on()
+    };
+
+    for parent in &parents {
+        if !has_diff(parent.tree()?)? {
+            return Ok(false);
+        }
     }
-    .block_on()
+
+    // Conflict resolution is expensive, try that only for matched files.
+    has_diff(rewrite::merge_commit_trees_no_resolve_without_repo(store, &index, &parents)?)
 }
 
 fn matches_diff_from_parent(

```

**Comment by yuja** [comment]
> Most savings come from not having to calculate the common ancestors for the merges.

Good find. The result isn't exactly the same, but it might be okay to approximate `diff(merge(parents), self) == false` as `all(diff(p, self) == false for p in parents)`.

> ```
> +    for parent in &parents {
> +        if !has_diff(parent.tree()?)? {
> +            return Ok(false);
> +        }
> ```

If one of the parents has difference, shouldn't we have to fall back to the slow `diff(merge(parents), self)` path?

**Comment by dotdash** [comment]
> > No, the idea is taken from how git does it (as we're comparing to git here).
> 
> Ah, okay. We've discussed that somewhere in this thread ([#4674 (comment)](https://github.com/jj-vcs/jj/issues/4674#issuecomment-2423476227) .. [#4674 (comment)](https://github.com/jj-vcs/jj/issues/4674#issuecomment-2490160968)), and afaict, we prefer to not introduce new git-like diff semantics.

I see. I did not get that impression from the thread. [Here](https://github.com/jj-vcs/jj/issues/4674#issuecomment-2424347932) Martin calls it a good idea, and I don't understand the [second comment](https://github.com/jj-vcs/jj/issues/4674#issuecomment-2490160968) that you linked as contradicting that. I only understood it to just refer to the fact that by default jj and git log may produce different amounts of output (by using `|head`), so timing them to reach a certain amount of output is unfair.

While trying a few things, I just noticed `jj` actually already omits merge commits if they're equal to one of the parents. So the above hack would possibly not even change any semantics and could just be applied, but I'm working from observations not code logic here. Example:

```❯ jj log -T builtin_log_oneline -r ..
@      lplruprs b.steinbrink 2025-05-27 13:45:33 5f89f66f merge delete
├─┬─╮
│ │ ○    lwlkxzws b.steinbrink 2025-05-27 13:44:13 master 0773e013 merge asd
│ │ ├─╮
│ ○ │ │  znxtwprk b.steinbrink 2025-05-27 13:44:47 0a41c2e2 merge bbb
│ ╰─┬─╮
○   │ │  zvqoyoqu b.steinbrink 2025-05-27 13:44:37 git_head() 6f5940f1 merge aaa
╰───┬─╮
    │ ○  tnpozxxv b.steinbrink 2025-05-27 13:44:01 a 2870ec7c aaa
    ○ │  qvsrrxwu b.steinbrink 2025-05-27 13:44:01 cfa44985 bbb
    ├─╯
    ○  nokkmvnz b.steinbrink 2025-05-27 13:43:58 6946112c   asd
    │
    ~
```
I created a file with content `asd`, two child commits with content changed to `aaa` and `bbb` respectively and then three merges, one with no changes relative to one of the parents (i.e. keeping either `aaa` or `bbb`, and one with changes relative to both parents changing the file back to `asd`. Finally a merge on top of those three that simply deletes the file.

Doing a log filtered by the file from the delete-merge ends up showing only one of its children (this might be a bug?):
```
❯ jj log -T builtin_log_oneline -r .. file
@  lplruprs b.steinbrink 2025-05-27 13:45:33 5f89f66f merge delete
○    lwlkxzws b.steinbrink 2025-05-27 13:44:13 master 0773e013 merge asd
├─╮
│ ○  tnpozxxv b.steinbrink 2025-05-27 13:44:01 a 2870ec7c aaa
○ │  qvsrrxwu b.steinbrink 2025-05-27 13:44:01 cfa44985 bbb
├─╯
○  nokkmvnz b.steinbrink 2025-05-27 13:43:58 6946112c   asd
│
~
```

Log starting at the `merge asd` commit gives the same as above minus the `merge delete` commit, but both logs for the `merge aaa` and `merge bbb` cases actually omit the merge commit in the about:
```
❯ jj log -T builtin_log_oneline -r ..zv file
○  qvsrrxwu b.steinbrink 2025-05-27 13:44:01 cfa44985 bbb
│ ○  tnpozxxv b.steinbrink 2025-05-27 13:44:01 a 2870ec7c aaa
├─╯
○  nokkmvnz b.steinbrink 2025-05-27 13:43:58 6946112c   asd
│
~

❯ jj log -T builtin_log_oneline -r ..zn file
○  qvsrrxwu b.steinbrink 2025-05-27 13:44:01 cfa44985 bbb
│ ○  tnpozxxv b.steinbrink 2025-05-27 13:44:01 a 2870ec7c aaa
├─╯
○  nokkmvnz b.steinbrink 2025-05-27 13:43:58 6946112c   asd
│
~
```

This is with an unmodified jj version`0.29.0-6cbb21f86f0347e6a1dacdeb5a15e04df8763664`

**Comment by dotdash** [comment]
Sure.
```bash
#!/usr/bin/env bash
jj git init --colocate       ; echo asd > file ; jj desc -m 'init asd'
jj new 'root()+'             ; echo aaa > file ; jj desc -m 'aaa'
jj new 'root()+'             ; echo bbb > file ; jj desc -m 'bbb'
jj new 'all:root()++'        ; echo aaa > file ; jj desc -m 'merge aaa'
jj new 'all:root()++'        ; echo bbb > file ; jj desc -m 'merge bbb'
jj new 'all:root()++'        ; echo asd > file ; jj desc -m 'merge asd'
jj new 'all:visible_heads()' ; rm file         ; jj desc -m 'merge delete'

```

**Comment by yuja** [comment]
Thanks, but the result looks good at my end.

```
% jj log -T builtin_log_oneline -r .. file
@      tksoqwlx yuya 2025-05-27 22:06:51 7a6cbb3f merge delete
├─┬─╮
│ │ o    vxulysxo yuya 2025-05-27 22:06:51 51698d11 merge aaa
│ │ ├─╮
│ o │ │  pwswuozt yuya 2025-05-27 22:06:51 4d3c0806 merge bbb
│ ╰─┬─╮
o   │ │  rrwqsroy yuya 2025-05-27 22:06:51 git_head() 46d3f78a merge asd
╰───┬─╮
    │ o  uxsqkvtk yuya 2025-05-27 22:06:51 b064160e aaa
    o │  tzmutwox yuya 2025-05-27 22:06:51 687e9e2e bbb
    ├─╯
    o  vwwvylvw yuya 2025-05-27 22:06:51 d9e2cd4e init asd
    │
    ~
% jj version
jj 0.29.0-aa81379534c90e41fda559e60c9799af370d752b
```

**Comment by martinvonz** [comment]
> [Here](https://github.com/jj-vcs/jj/issues/4674#issuecomment-2424347932) Martin calls it a good idea

Sorry, I take that back :) It seems like it should rarely help (it only helps if the changes from one side were discarded, right?), but I heard that you said that it does make a big difference in practice. I would prefer improvements that don't change the semantics.

**Comment by dotdash** [comment]
FWIW, Yuya's recent [optimization](https://github.com/jj-vcs/jj/pull/6650) to the code that determines the common ancestors of merge commits seems to make the whole code path as fast as my hack in has_diff(), and it probably (didn't test yet) also improves the performance of the `empty` template in the log output. Great work!

**Comment by arianvp** [comment]
This doesn't seem fixed to me at all.

```
jj git clone https://github.com/nixos/nixpkgs
jj log ./ci/pinned.json
```

This hangs already for over 5 minutes. I don't think it'll complete.

whilst `git log ./ci/pinned.json` is instant

**Comment by PhilipMetzger** [comment]
Since `Nixpkgs` is a merge heavy repo, you need to `jj debug index-changed-paths` to make `jj log` faster.
